File: backend/app/models/ab_testing.py
"""
A/B Testing Models for Agent Evolution

Tracks experiments, variants, and results for continuous evolution.
"""
from datetime import datetime
from typing import Optional
from sqlalchemy import Column, String, Integer, Float, Boolean, DateTime, ForeignKey, Text, CheckConstraint, Index
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
import uuid

from app.database import Base


class Experiment(Base):
    """
    An A/B test experiment for an agent feature.

    Lifecycle: DRAFT -> ACTIVE -> COMPLETED/CANCELLED
    Can be PAUSED temporarily.
    """
    __tablename__ = "experiments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_id = Column(String(50), nullable=False, index=True)
    feature_name = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)

    # Status with CHECK constraint
    status = Column(String(20), nullable=False, default="DRAFT")

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)

    # Winner tracking
    winner_variant_id = Column(UUID(as_uuid=True), ForeignKey('experiment_variants.id'), nullable=True)

    # Extra metadata (renamed from 'metadata' to avoid SQLAlchemy reserved name)
    extra_metadata = Column(JSONB, nullable=True, default={})

    # Relationships
    variants = relationship("ExperimentVariant", back_populates="experiment", foreign_keys="ExperimentVariant.experiment_id")
    winner_variant = relationship("ExperimentVariant", foreign_keys=[winner_variant_id], post_update=True)

    # Constraints
    __table_args__ = (
        CheckConstraint(
            "status IN ('DRAFT', 'ACTIVE', 'PAUSED', 'COMPLETED', 'CANCELLED')",
            name='check_experiment_status'
        ),
        Index('idx_experiment_agent_status', 'agent_id', 'status'),
    )

    def __repr__(self):
        return f"<Experiment {self.id} agent={self.agent_id} feature={self.feature_name} status={self.status}>"


class ExperimentVariant(Base):
    """
    A variant in an A/B test experiment.

    Contains configuration and traffic allocation percentage.
    """
    __tablename__ = "experiment_variants"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    experiment_id = Column(UUID(as_uuid=True), ForeignKey('experiments.id', ondelete='CASCADE'), nullable=False, index=True)

    variant_name = Column(String(100), nullable=False)  # e.g., "control", "variant_a"
    description = Column(Text, nullable=True)

    # Traffic allocation (must sum to 100 across variants)
    traffic_percentage = Column(Float, nullable=False)

    # Configuration JSON (agent-specific config)
    config_json = Column(JSONB, nullable=False, default={})

    # Is this the control/baseline variant?
    is_control = Column(Boolean, nullable=False, default=False)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    # Relationships
    experiment = relationship("Experiment", back_populates="variants", foreign_keys=[experiment_id])
    results = relationship("ExperimentResult", back_populates="variant", cascade="all, delete-orphan")

    # Constraints
    __table_args__ = (
        CheckConstraint('traffic_percentage >= 0 AND traffic_percentage <= 100', name='check_traffic_percentage'),
        Index('idx_variant_experiment', 'experiment_id'),
    )

    def __repr__(self):
        return f"<ExperimentVariant {self.id} name={self.variant_name} traffic={self.traffic_percentage}%>"


class ExperimentResult(Base):
    """
    A single result/outcome from an experiment variant.

    Tracks task execution success and metrics.
    """
    __tablename__ = "experiment_results"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    variant_id = Column(UUID(as_uuid=True), ForeignKey('experiment_variants.id', ondelete='CASCADE'), nullable=False, index=True)

    # Task tracking
    task_id = Column(String(100), nullable=False, index=True)
    work_type = Column(String(50), nullable=True)

    # Outcome
    success = Column(Boolean, nullable=False)

    # Metrics (JSON with flexible schema)
    metrics_json = Column(JSONB, nullable=False, default={})
    # Example metrics:
    # {
    #   "success_rate": 0.95,
    #   "estimation_accuracy": 0.85,
    #   "code_quality_score": 8.5,
    #   "execution_time_ms": 1500,
    #   "retry_count": 1
    # }

    # Error tracking (if success=False)
    error_message = Column(Text, nullable=True)

    # Timestamp
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)

    # Relationships
    variant = relationship("ExperimentVariant", back_populates="results")

    # Indexes
    __table_args__ = (
        Index('idx_result_variant_created', 'variant_id', 'created_at'),
        Index('idx_result_task', 'task_id'),
    )

    def __repr__(self):
        return f"<ExperimentResult {self.id} variant={self.variant_id} success={self.success}>"

File: agents/types/ABTesting.ts
/**
 * A/B Testing Types for Agent Evolution
 *
 * Multi-variant experimentation framework for continuous agent improvement.
 */

/**
 * Experiment status enum
 */
export enum ExperimentStatus {
    DRAFT = 'DRAFT',           // Setup phase, can edit variants
    ACTIVE = 'ACTIVE',         // Running, traffic allocated
    PAUSED = 'PAUSED',         // Temporarily stopped
    COMPLETED = 'COMPLETED',    // Winner declared
    CANCELLED = 'CANCELLED'     // Aborted without winner
}

/**
 * Traffic allocation configuration
 */
export interface TrafficAllocation {
    variantId: string;
    percentage: number;  // 0-100
}

/**
 * Metrics snapshot for experiment result
 */
export interface MetricsSnapshot {
    success_rate?: number;           // 0-1
    estimation_accuracy?: number;    // 0-1
    code_quality_score?: number;     // 0-10
    execution_time_ms?: number;
    retry_count?: number;
    [key: string]: any;  // Allow additional metrics
}

/**
 * Experiment variant configuration
 */
export interface ExperimentVariant {
    id: string;
    experiment_id: string;
    variant_name: string;           // e.g., "control", "variant_a"
    description?: string;
    traffic_percentage: number;      // 0-100
    config_json: Record<string, any>; // Agent-specific config
    is_control: boolean;
    created_at: Date;
}

/**
 * A/B test experiment
 */
export interface Experiment {
    id: string;
    agent_id: string;               // Which agent is being tested
    feature_name: string;           // Feature being experimented with
    description?: string;
    status: ExperimentStatus;
    created_at: Date;
    started_at?: Date;
    completed_at?: Date;
    winner_variant_id?: string;
    extra_metadata?: Record<string, any>;

    // Relationships (populated)
    variants?: ExperimentVariant[];
    winner_variant?: ExperimentVariant;
}

/**
 * Single result from a variant
 */
export interface ExperimentResult {
    id: string;
    variant_id: string;
    task_id: string;
    work_type?: string;
    success: boolean;
    metrics_json: MetricsSnapshot;
    error_message?: string;
    created_at: Date;

    // Relationship (populated)
    variant?: ExperimentVariant;
}

/**
 * Statistical analysis result
 */
export interface ExperimentAnalysis {
    experiment_id: string;
    sample_size_per_variant: Record<string, number>;
    success_rate_per_variant: Record<string, number>;
    confidence_intervals: Record<string, { lower: number; upper: number }>;
    p_value?: number;
    recommended_winner?: string;
    consensus_level: number;  // 0-1 (confidence in winner)
    min_samples_reached: boolean;
}

/**
 * Request to create new experiment
 */
export interface CreateExperimentRequest {
    agent_id: string;
    feature_name: string;
    description?: string;
    variants: CreateVariantRequest[];
}

/**
 * Request to create variant
 */
export interface CreateVariantRequest {
    variant_name: string;
    description?: string;
    traffic_percentage: number;
    config_json: Record<string, any>;
    is_control?: boolean;
}

/**
 * Request to start experiment
 */
export interface StartExperimentRequest {
    experiment_id: string;
}

/**
 * Request to complete experiment
 */
export interface CompleteExperimentRequest {
    experiment_id: string;
    winner_variant_id: string;
}

/**
 * Request to log experiment result
 */
export interface LogResultRequest {
    variant_id: string;
    task_id: string;
    work_type?: string;
    success: boolean;
    metrics: MetricsSnapshot;
    error_message?: string;
}

/**
 * Helper: Calculate traffic allocation for task
 */
export function allocateTraffic(
    variants: ExperimentVariant[],
    taskId: string,
    experimentId: string
): string {
    // Deterministic hash for sticky assignment
    const hash = simpleHash(`${taskId}-${experimentId}`);
    const randomValue = hash % 100;  // 0-99

    let cumulative = 0;
    for (const variant of variants) {
        cumulative += variant.traffic_percentage;
        if (randomValue < cumulative) {
            return variant.id;
        }
    }

    // Fallback to first variant
    return variants[0]?.id || '';
}

/**
 * Simple hash function for consistent assignment
 */
function simpleHash(str: string): number {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        const char = str.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash;  // Convert to 32bit integer
    }
    return Math.abs(hash);
}

/**
 * Validation: Check if traffic percentages sum to 100
 */
export function validateTrafficAllocation(
    variants: CreateVariantRequest[]
): boolean {
    const total = variants.reduce((sum, v) => sum + v.traffic_percentage, 0);
    return Math.abs(total - 100) < 0.01;  // Allow for floating point errors
}

/**
 * Helper: Check if minimum samples reached for statistical significance
 */
export function hasMinimumSamples(
    resultsPerVariant: Record<string, number>,
    minSamples: number = 30
): boolean {
    return Object.values(resultsPerVariant).every(count => count >= minSamples);
}

/**
 * Helper: Calculate success rate from results
 */
export function calculateSuccessRate(results: ExperimentResult[]): number {
    if (results.length === 0) return 0;
    const successes = results.filter(r => r.success).length;
    return successes / results.length;
}

/**
 * Helper: Group results by variant
 */
export function groupResultsByVariant(
    results: ExperimentResult[]
): Record<string, ExperimentResult[]> {
    return results.reduce((acc, result) => {
        if (!acc[result.variant_id]) {
            acc[result.variant_id] = [];
        }
        acc[result.variant_id].push(result);
        return acc;
    }, {} as Record<string, ExperimentResult[]>);
}

/**
 * Export all types
 */
export type {
    TrafficAllocation,
    MetricsSnapshot,
    ExperimentVariant,
    Experiment,
    ExperimentResult,
    ExperimentAnalysis,
    CreateExperimentRequest,
    CreateVariantRequest,
    StartExperimentRequest,
    CompleteExperimentRequest,
    LogResultRequest
};
